File: src/Systems/Objective-C/composite.m
#import <Foundation/Foundation.h>

@protocol GKComponent <NSObject>
- (NSInteger)size;
@end

@interface GKFileLeaf : NSObject <GKComponent>
- (instancetype)initWithBytes:(NSInteger)bytes;
@end

@implementation GKFileLeaf {
    NSInteger _bytes;
}

- (instancetype)initWithBytes:(NSInteger)bytes {
    self = [super init];
    if (self) {
        _bytes = bytes;
    }
    return self;
}

- (NSInteger)size {
    return _bytes;
}
@end

@interface GKFolderComposite : NSObject <GKComponent>
- (instancetype)initWithChildren:(NSArray<id<GKComponent>> *)children;
@end

@implementation GKFolderComposite {
    NSArray<id<GKComponent>> *_children;
}

- (instancetype)initWithChildren:(NSArray<id<GKComponent>> *)children {
    self = [super init];
    if (self) {
        _children = [children copy];
    }
    return self;
}

- (NSInteger)size {
    NSInteger total = 0;
    for (id<GKComponent> child in _children) {
        total += [child size];
    }
    return total;
}
@end

int main(void) {
    @autoreleasepool {
        id<GKComponent> readme = [[GKFileLeaf alloc] initWithBytes:2];
        id<GKComponent> docs = [[GKFolderComposite alloc] initWithChildren:@[
            [[GKFileLeaf alloc] initWithBytes:3],
            [[GKFileLeaf alloc] initWithBytes:5]
        ]];
        id<GKComponent> root = [[GKFolderComposite alloc] initWithChildren:@[readme, docs]];

        printf("leaf=%ld\n", (long)[readme size]);
        printf("docs=%ld\n", (long)[docs size]);
        printf("root=%ld\n", (long)[root size]);
    }
    return 0;
}

File: src/Scripting/Perl/composite.pl
use strict;
use warnings;

package FileLeaf;
sub new {
    my ($class, $bytes) = @_;
    return bless { bytes => $bytes }, $class;
}
sub size { return $_[0]->{bytes}; }

package FolderComposite;
sub new {
    my ($class, @children) = @_;
    return bless { children => \@children }, $class;
}
sub size {
    my ($self) = @_;
    my $total = 0;
    $total += $_->size() for @{$self->{children}};
    return $total;
}

package main;
my $readme = FileLeaf->new(2);
my $docs = FolderComposite->new(FileLeaf->new(3), FileLeaf->new(5));
my $root = FolderComposite->new($readme, $docs);

print "leaf=" . $readme->size() . "\n";
print "docs=" . $docs->size() . "\n";
print "root=" . $root->size() . "\n";
